### PR TITLE
Vectorize `archivePackages()` and `pruneRepo()`

### PR DESCRIPTION
**Pull request by pat-s**
Besides vectorizing both functions, I've also added the ability to archive packages for arbitrary R versions for every instance of "type".

By default (version = NULL), existing R versions within the given repopaths are checked and used.

Hence, by default all available "type"s are executed with all existing R version found in their respective directories.

I've already tested the approach on mlr3learners.drat in which we have 21 packages with binaries for different R versions.
So far it works fine.

I added two calls to {stringr} which could probably replaced by base R functions. If that's a blocker I can get rid of them. However, extracting substr with base R is not so nice because substr() does not support regexs.

If you want to try it out, you can do the following locally while having this branch installed (`remotes::install_github("pat-s/drat@vectorize-archive")`):

```
git clone git@github.com:mlr3learners/mlr3learners.drat.git
rm -rf bin/
rm -rf src/
git reset --hard 6c4ba1625f337e5dcf733eebd9fd0a7a73f071f8
```

and then call `drat::archivePackages(".")`.

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -1,14 +1,14 @@
 Package: drat
 Type: Package
 Title: 'Drat' R Archive Template
-Version: 0.1.6
-Date: 2020-05-29
+Version: 0.2.0
```
Comment (eddelbuettel) [inline DESCRIPTION:4]: Please leave changes to version and dates to the release manager aka maintainer.  Thank you.

File: DESCRIPTION [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
+Date: 2020-06-05
 Author: Dirk Eddelbuettel with contributions by Carl Boettiger, Neal Fultz,
  Sebastian Gibb, Colin Gillespie, Jan Górecki, Matt Jones, Thomas Leeper,
- Steven Pav, Jan Schulz, Christoph Stepper and Felix G.M. Ernst. 
+ Steven Pav, Jan Schulz, Christoph Stepper, Felix G.M. Ernst and Patrick Schratz.
```
Comment (eddelbuettel) [inline DESCRIPTION:8]: Idem.

File: DESCRIPTION [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
+Version: 0.2.0
+Date: 2020-06-05
 Author: Dirk Eddelbuettel with contributions by Carl Boettiger, Neal Fultz,
  Sebastian Gibb, Colin Gillespie, Jan Górecki, Matt Jones, Thomas Leeper,
- Steven Pav, Jan Schulz, Christoph Stepper and Felix G.M. Ernst. 
+ Steven Pav, Jan Schulz, Christoph Stepper, Felix G.M. Ernst and Patrick Schratz. 
 Maintainer: Dirk Eddelbuettel <edd@debian.org>
 Depends: R (>= 3.2.0)
-Imports: utils
+Imports: utils, stringr
```
Comment (eddelbuettel) [inline DESCRIPTION:11]: We need clear motivation for expanding dependencies.  This does not qualify.

File: NAMESPACE [hunk 4]
```diff
@@ -1,5 +1,6 @@
 importFrom(tools, "write_PACKAGES")
 importFrom(utils, "untar", "unzip", "contrib.url")
+importFrom(stringr, "str_extract")
```
Comment (eddelbuettel) [inline NAMESPACE:3]: Same.

File: R/archivePackages.R [hunk 5]
```diff
@@ -1,67 +1,93 @@
...
-##' The function moves older versions of packages into a CRAN-style 
+##' The function moves older versions of packages into a CRAN-style
 ##' archive folder.
-##' 
+##'
 ##' This function is still undergoing development and polish and may
 ##' change in subsequent versions.
 ##'
 ##' @title Move older copies of packages to an archive
+##' @importFrom stringr str_extract
```
Comment (eddelbuettel) [inline R/archivePackages.R:9]: Please remove / replace.

File: R/archivePackages.R [hunk 6]
```diff
@@ -1,67 +1,93 @@
...
-    if (!file.exists(archive)) {
-        if (!dir.create(archive, recursive = TRUE)) {
-            stop("Archive directory not found and couldn't be created\n", call. = FALSE)
+    for (type in type) {
+        
+        # if no R version is supplied, the one from the
+        # respective "type" directory is scraped
+        if (is.null(version)) {
+            if (grepl("binary", type)) {
+                r_versions <- stringr::str_extract(list.dirs(gsub(
```
Comment (eddelbuettel) [inline R/archivePackages.R:43]: I appreciate your offer to rewrite this and not use extra dependencies for a function that is a bit more tangential to the package.   

And if and when you rewite please avoid excess line breaks.   100 cols are fine.    You can see in the rest of the file that we often use close to 100 cols (and the one at 111 cols could be considered too long ...  but it works so no need for a PR or change of working code).

File: DESCRIPTION (hunk 3 above)
Comment (jangorecki) [reply DESCRIPTION:11]: `stringr` is a heavy dependency, such change would make `drat` unusable on many lightweight configurations.

File: R/archivePackages.R [hunk 7]
```diff
@@ -10,58 +10,84 @@
...
 archivePackages <- function(repopath = getOption("dratRepo", "~/git/drat"),
-                            type = "source", 
-                            pkg,
-                            version = getRversion()) {
+                            type = c(
+                                "source", "mac.binary", "mac.binary.el-capitan",
+                                "mac.binary.mavericks", "win.binary"
+                            ),
+                            pkg = NULL,
+                            version = NULL) {
```
Comment (eddelbuettel) [inline R/archivePackages.R:34]: The suggested change is now simpler and more focussed.  One thing some people have learned to not be cavalier about is the change if existing interfaces.   `pkg` and `version` defaulted to "possibly empty" with a test for `is.missing`.  Having `NULL` is similar but not the same, and it *is* different.   There may be users out there with an existing script calling this (I am actually not entirely sure if/how anybody uses `archivePackages()` but I cannot know).  The use case of 'called with no argument' is preserved so I guess I can call that good enough.

File: R/archivePackages.R [hunk 8]
```diff
@@ -10,58 +10,84 @@
...
+            } else {
+                # no version for source packages
+                r_versions <- ""
+            }
+        } else {
+            dirs = list.dirs(gsub(
+                "/[0-9].[0-9]$", "", contrib.url(repopath, type)), 
+                recursive = FALSE)
+            r_versions = substring(dirs, regexpr("[0-9].[0-9]$", dirs))
         }
```
Comment (eddelbuettel) [inline R/archivePackages.R:55]: Still not my style of writing but then I just like others think how R Core writes code matters most, and i happen to use the same editor they use (or at least a number of them).   So whatevs.   Thanks for adhering to four space indentation though, that is appreciated.  (And this repo is so small that I never bothered with `,editoconfig` which may be overkill.)

File: R/archivePackages.R [hunk 9]
```diff
@@ -10,58 +10,84 @@
...
-        pkg <- unique(pkg)
-        old <- pruneRepo(repopath = repopath, type = type, pkg = pkg,
-                         version = version, remove = FALSE)
-        old <- old[!old[,"newest"] & old[,"package"] %in% pkg, ]
-        sapply(pkg, mkArchive)
-    }
-    file.rename(file.path(repodir, old$file), file.path(repodir, "Archive", old$package, old$file))
-    
-    invisible(NULL)
+    return(invisible(NULL))
```
Comment (eddelbuettel) [inline R/archivePackages.R:92]: Adding invibsible is good style now that df may be larger.

**Comment by eddelbuettel** [comment]
@jangorecki Any comments?

**Comment by eddelbuettel** [comment]
Also, and that is of course a little out there and not a "formal" requirement:  how would you go about testing this?   You  told use several times already that you use it for mlr3 repo or subrepo so is the best answer "in prod" ?

**Comment by eddelbuettel** [comment]
The PR is also a roxygen run behind.  I just added that.

**Comment by eddelbuettel** [comment]
And just for completeness, the [littler script dratInsert.r](https://github.com/eddelbuettel/littler/blob/master/inst/examples/dratInsert.r) still works with this code and just up a new [Rcpp dev version 1.0.4.12](https://github.com/RcppCore/drat/commit/09c62c79c7055464b43e1e3b09719649c909c0e5) so the core functionality is still there too...

**Comment by pat-s** [comment]
> Also, and that is of course a little out there and not a "formal" requirement: how would you go about testing this? You told use several times already that you use it for mlr3 repo or subrepo so is the best answer "in prod" ?

Yes, and given the complexity in testing this, it is sufficient for me here.

One could of course replicate a small drat structure on the fly during tests and then reset it's state at the end of the tests. However, testing the results automatically is also not that easy. 
Looking at the results in practice/production after triggering the function works.
The code snippet I posted should also work in the future since the repo won't go away.
Also there is only few risk that the initial repo structure is in a different state than what it should be (possibly causing the function to fail).

---

Feel free to make additional changes, including indention/coding style.

**Comment by eddelbuettel** [comment]
@FelixErnst This PR changes some of the code you added recently and extends it.  As you, if I remember correctly, use this "in prod" could you take a look and (if you have the time) a quick test drive of the branch / pr ?

File: R/archivePackages.R [hunk 10]
```diff
@@ -10,58 +10,84 @@
...
-        if (!dir.create(archive, recursive = TRUE)) {
-            stop("Archive directory not found and couldn't be created\n", call. = FALSE)
+    for (type in type) {
+        
+        # if no R version is supplied, the one from the
+        # respective "type" directory is scraped
+        if (is.null(version)) {
+            if (grepl("binary", type)) {
+                dirs = list.dirs(gsub(
+                    "/[0-9].[0-9]$", "", contrib.url(repopath, type)),
```
Comment (FelixErnst) [inline R/archivePackages.R:43]: I am not sure, but this regex does not masks `.` It works here, but I think `/336` would also match. No problem here, but `/[0-9]\\.[0-9]$` would be better

**Comment by pat-s** [comment]
Felix,

lot's can be made more sophisticated here, one could always add more assertions or tests.
This is a bare-bone version of extending functionality and making things easier for the user at the same time. The "bare-bone state" of most functions in this package before this PR was not different than with this new addition.

It works for the job that is should do and I also do not see much danger in wrong arguments inputs from users as the usage is pretty much tied to a properly structured drat repo anyways.

I do not have more time budget to refine this PR substantially. Happy to give you push access to the fork if you want to do so.

> edit2: the code in this PR works as intended and solves the issue

I'm not sure what this means here - does it invalidate some of your previous comments?

**Comment by FelixErnst** [comment]
Hi Patrick @pat-s,

The code in your PR, works as intended. So my comments are just about the logic in the code, which I think need to be improved, so that the code can be modified in incremental steps from this point forward, so that user contributions can be made easily.

In R, in my opinion, and this just me talking, if you use the broad stroke functions `for` or `lapply` it has the danger of putting everything in a jacket and thus, hinders small step development.

I would be very happy to push the things I have in mind to your fork, if you like.

File: R/archivePackages.R [hunk 11]
```diff
@@ -1,93 +1,128 @@
...
+    } else {
+        pkg <- unique(pkg)
+    }
+    pkg
+}
+
+##' The function moves older versions of packages into a CRAN-style 
 ##' archive folder.
-##'
+##'
```
Comment (jangorecki) [inline R/archivePackages.R]: Seeing one commit to remove empty trailing space, and soon after another commit to introduce it is kind of funny.
I suggest to always check `git diff --cached` before doing `commit` to ensure that lines which are affected are the lines necessary for logic change. If one wants to change existing style/formattng, then it is always better to put into separate commit, or even better, not do it at all. The smaller the diff, the better, and easier to review. This commit did actually very well by removing previously affected lines.

Comment (eddelbuettel) [reply R/archivePackages.R]: Not all editors are equally smart. Adding `.editorconfig` helps but not all editors adhere to it...

Comment (FelixErnst) [reply R/archivePackages.R]: I maybe added that. I know it says outdated but I had to add the `@name` tag and RStudio behaves a bit differently, when automatically adding new lines in the comment section.

File: R/pruneRepo.R [hunk 12]
```diff
@@ -35,93 +35,108 @@
...
+            "\\.tgz$"
+        } else if (x == "mac.binary.mavericks") {
+            "\\.tgz$"
+        } else if (x == "win.binary") {
+            "\\.zip$"
         } else {
-            for (f in rmfiles) {
-                fullfile <- file.path(repodir, f)
-                unlink(fullfile)
+            stop("Unknown package type. Valid values are 'source', 'mac.binary', 'mac.binary.el-capitan', 'mac.binary.mavericks' and 'win.binary'.", call. = FALSE)
```
[the remark this answers is not in the document]
Comment (FelixErnst) [reply R/pruneRepo.R]: However, `binary` is not specific. `binary` means in this case all `binary` cases, `win.binary`, `mac.binary` and so on.

A logic for this is now implemented including also the broadest term `both`.

File: R/insertPackage.R [hunk 13]
```diff
@@ -159,10 +159,20 @@ insertPackage <- function(file,
...
+insertPackages <- function(files, ...){
+    lapply(files, insertPackage, ...)
+}
 
 ##' @rdname insertPackage
-insert <- function(...) insertPackage(...)
+insert <- function(...) {
+    args <- list(...)
+    if("files" %in% names(args)){
+        insertPackages(...)
```
Comment (jangorecki) [inline R/insertPackage.R:171]: how named `files` argument to `...` is meant to work here? are there any examples or documentation?

File: R/insertPackage.R [hunk 14]
```diff
@@ -246,14 +256,31 @@ getPackageInfo <- function(file) {
...
+        contrib_url
     }
-    version <- package_version(version)
-    contrib_url <- gsub("contrib/[0-9]\\.[0-9]",
-                        paste0("contrib/",
-                               paste0(version$major,".",version$minor)),
-                        contrib_url)
-    contrib_url
+    urls <- lapply(type,FUN)
+    names <- unlist(mapply(rep,type,lengths(urls)))
```
Comment (jangorecki) [inline R/insertPackage.R:282]: I haven't investigate inputs here, but if you `unlist` here anyway, then safer to use `mapply(SIMPLIFY=FALSE)`

File: R/pruneRepo.R [hunk 15]
```diff
@@ -56,36 +74,49 @@ pruneRepo <- function(repopath = getOption("dratRepo", "~/git/drat"),
     } else {
         stop("Unknown package type", call. = FALSE)
     }
+    ext
+}
+
+.get_repoinfo_per_package_type <- function(t, e, rd, pkg){
+    files <- list.files(rd, pattern = e, full.names = FALSE)
+    if(length(files) == 0L){
+        return(data.frame())
```
Comment (jangorecki) [inline R/pruneRepo.R:83]: maybe better 0 rows data.frame but still providing empty columns? then less exception to handle when processing output of this function...

File: R/pruneRepo.R [hunk 16]
```diff
@@ -98,30 +129,55 @@ pruneRepo <- function(repopath = getOption("dratRepo", "~/git/drat"),
     ## 8   RcppTOML_0.0.3.tar.gz  RcppTOML    0.0.3   TRUE
     ## 9  winsorize_0.0.2.tar.gz winsorize    0.0.2   TRUE
     ## R>
+    df
+}
 
-    haspkg <- requireNamespace("git2r", quietly = TRUE)
-    
+##' @rdname pruneRepo
+pruneRepo <- function(repopath = getOption("dratRepo", "~/git/drat"),
```
Comment (jangorecki) [inline R/pruneRepo.R:136]: shouldn't this be `file.path()` rather than `/` ?

File: R/insertPackage.R (hunk 13 above)
Comment (FelixErnst) [reply R/insertPackage.R:171]: See line 31 in this file

File: R/pruneRepo.R (hunk 15 above)
Comment (FelixErnst) [reply R/pruneRepo.R:83]: It is filtered out by using 'nrow() == 0L'. See line 58. Could also be done by returning `NULL` in this case, but other cases of zero-row data.frames exist in that function as well. This solution takes care of all of them.

File: R/pruneRepo.R (hunk 16 above)
Comment (FelixErnst) [reply R/pruneRepo.R:136]: Not sure what you mean. The first argument of `pruneRepo` was not changed.

File: R/insertPackage.R (hunk 14 above)
Comment (FelixErnst) [reply R/insertPackage.R]: fixed in 8a15269

File: R/insertPackage.R (hunk 13 above)
Comment (jangorecki) [reply R/insertPackage.R:171]: I think providing vector input to `file` argument is fine too, shouldn't need extra `files` argument handling

File: R/pruneRepo.R (hunk 15 above)
Comment (jangorecki) [reply R/pruneRepo.R:83]: ok, it is internal is should be fine

File: R/pruneRepo.R (hunk 16 above)
Comment (jangorecki) [reply R/pruneRepo.R:136]: good then

File: R/insertPackage.R (hunk 13 above)
Comment (FelixErnst) [reply R/insertPackage.R]: fixed in 207fa40

**Comment by FelixErnst** [comment]
I added fixed test packages in `extdata` for macOS `3.6` and `4.0`, windows `4.0` and source with version `1.0`, `1.1` and `1.2` each. I hope that is ok. It adds 100Kb to the package, but the tests did unearth some small fixes needed.

**Comment by eddelbuettel** [comment]
@FelixErnst  That is pretty awesome -- but it also _broke the tests_:

image

Can you please look into restoring the package to passing tests. either by correcting what is currently broken (preferred) or removing broken tests (less ideal)?

Also, can we slow down a little and maybe discuss what may get committed next and get agreement by all (or most) before just going off and doing it?   We're almost there here and I really appreciate the input but the repo has gotten a little hyperactive.

**Comment by FelixErnst** [comment]
> @FelixErnst That is pretty awesome -- but it also _broke the tests_:
>
> Can you please look into restoring the package to passing tests. either by correcting what is currently broken (preferred) or removing broken tests (less ideal)?

The error occurs because a certain version of the `.Rbuildignore` file is used on travis-CI. The last line `^.*\.tar\.gz$` is the problem and this was corrected on this branch, but somehow persists for the travis-CI build. I don't know, how I could fix this for travis-CI. (In addition, I don't know, why it was there in the first place. Maybe I overlooked something)

> Also, can we slow down a little and maybe discuss what may get committed next and get agreement by all (or most) before just going off and doing it? We're almost there here and I really appreciate the input but the repo has gotten a little hyperactive.

Of course. I had the time and I was on a roll and really like the issues raised by @pat-s and the additional functionality if offers. I am done with the restructuring I had in mind.

**Comment by eddelbuettel** [comment]
I see.  The Travis setup uses the simple `run.sh` from [this repo](https://github.com/eddelbuettel/r-travis), notably the version in [this file from docs/run.sh](https://github.com/eddelbuettel/r-travis/blob/master/docs/run.sh).  I was worried it may tamper with that expression, but it doesn't. It does however add three items (if missing) on [lines 65 to 73](https://github.com/eddelbuettel/r-travis/blob/master/docs/run.sh#L65-L73).  Could it be as simple as a missing newline?

You can always cat a `cat .Rbuildignore` to `.travis.yml` to see what it is like...

**Comment by eddelbuettel** [comment]
Slight thinko on my part.  The existing `^.*\.tar\.gz$` is in there because I tend to build `tar.gz` with the a simpler littler script that leaves them in the repo directory (to not clutter the parent directory).  But one then needs to exclude the previous source `tar.gz` from the current one being built.  So here maybe we want `drat_.*\.tar\.gz$` instead as the final line.  And ... make sure a new line gets added.  I should probably robustify `run.sh` for that too.

**Comment by FelixErnst** [comment]
hmm. the error persists. Can one get a hold of the status of the virtual machine from travis-CI?

I encountered the same problem locally and fixed it, but without investigating what is going one travis-CI, I don't have any idea what might be going on here.

I can of course remove the tests...

**Comment by eddelbuettel** [comment]
I think we should remove the tests and table them for a later addition and get on with our lifes. 

I think you can come close to mimicking Travis CI by starting in a Ubuntu 18.04 "bionic" container, but it is not perfect.  (Which is one of the reasons I went to fully Dockerized CI tests where Travis CI simply calls the Docker container I provide -- much better control over what is present and how. Overkill for `drat`, of course).

File: .Rbuildignore [hunk 17]
```diff
@@ -3,4 +3,4 @@
 .travis.yml
 vignettes/Makefile
 testpackages
-^.*\.tar\.gz$
\ No newline at end of file
+drat_.*\.tar\.gz$
```
Comment (eddelbuettel) [inline .Rbuildignore:6]: Change may now not be needed but whatevs.

File: DESCRIPTION [hunk 18]
```diff
@@ -1,11 +1,12 @@
 Package: drat
 Type: Package
 Title: 'Drat' R Archive Template
-Version: 0.1.6
-Date: 2020-05-29
+Version: 0.1.6.4
+Date: 2020-06-09
```
Comment (eddelbuettel) [inline DESCRIPTION:5]: I told @pat-s for "mucking" with Version and Date as that is what the release manager aka maintainer does.  But at least you used the pattern I use which is a minimal delta increment (whereas he forcefully suggested that his changes should bump immediately to 0.2.0 which is .... something we agreed to reverse).   This is fine.

File: DESCRIPTION [hunk 19]
```diff
@@ -1,11 +1,12 @@
...
 Author: Dirk Eddelbuettel with contributions by Carl Boettiger, Neal Fultz,
  Sebastian Gibb, Colin Gillespie, Jan Górecki, Matt Jones, Thomas Leeper,
- Steven Pav, Jan Schulz, Christoph Stepper and Felix G.M. Ernst. 
+ Steven Pav, Jan Schulz, Christoph Stepper, Felix G.M. Ernst and Patrick 
+ Schratz.
```
Comment (eddelbuettel) [inline DESCRIPTION:9]: Similar. My project. At the end of the day shouldn't I decide whom I anoint as a contributor?  No need to change now, but please consider it next time.

File: NAMESPACE [hunk 20]
```diff
@@ -3,5 +3,9 @@ importFrom(utils, "untar", "unzip", "contrib.url")
 export(addRepo,
        initRepo,
        insertPackage,
+       insertPackages,
+       getRepoInfo,
        pruneRepo,
-       archivePackages)
+       pruneRepoForAllRversions,
+       archivePackages,
+       archivePackagesForAllRversions)
```
Comment (eddelbuettel) [inline NAMESPACE:11]: I really want to be done with the PR but dropping new functions into the API without first discussing it is ... a little on the strong-willed side too.  Here I am ok with your changes; the process was a little on the heavy side.

File: drat.Rproj [hunk 21]
```diff
@@ -1,16 +0,0 @@
...
-EnableCodeIndexing: Yes
-UseSpacesForTab: Yes
-NumSpacesForTab: 2
-Encoding: UTF-8
-
-RnwWeave: knitr
-LaTeX: pdfLaTeX
-
-BuildType: Package
-PackageInstallArgs: --no-multiarch --with-keep.source
```
Comment (eddelbuettel) [inline drat.Rproj:16]: Why was this file deleted?

**Comment by eddelbuettel** [comment]
Lastly, I really like the code changes you brought.  They help.

We had bad luck with the test, and had to revert. Stuff happens.  Now that we reverted, should these not be deleted?

```sh
edd@rob:~/git/drat(vectorize-archive)$ tree inst/extdata/
inst/extdata/
├── 3.6
│   ├── foo_1.0.tgz
│   ├── foo_1.1.tgz
│   └── foo_1.2.tgz
├── 4.0
│   ├── foo_1.0.tgz
│   ├── foo_1.0.zip
│   ├── foo_1.1.tgz
│   ├── foo_1.1.zip
│   ├── foo_1.2.tgz
│   └── foo_1.2.zip
└── src
    ├── foo_1.0.tar.gz
    ├── foo_1.1.tar.gz
    └── foo_1.2.tar.gz

3 directories, 12 files
edd@rob:~/git/drat(vectorize-archive)$ 
```

File: NAMESPACE (hunk 20 above)
Comment (FelixErnst) [reply NAMESPACE:11]: Sorry for that. I didn't want to interfere with the expectation of current users, that more than R version would be targeted. I couldn't think of an alternative.

File: drat.Rproj (hunk 21 above)
Comment (FelixErnst) [reply drat.Rproj:16]: Because I made a mistake.

**Comment by FelixErnst** [comment]
Nope. The CI problem was two fold.

1. I didn't check my `git add` carefully enough. a `.gitignore` statement interefered with `.tar.gz` files being added (Hence, "the missing files... -_-" commit)
2. The .Rbuildignore removed them from the check dir.

The test work now as intended and the files list above are used for testing.

File: drat.Rproj (hunk 21 above)
Comment (eddelbuettel) [reply drat.Rproj:16]: Mistakes happen. I think it was mostly auto-generated. I can also make a copy of the one I have here (pre-merge).

File: NAMESPACE (hunk 20 above)
Comment (eddelbuettel) [reply NAMESPACE:11]: I think the design is entirely defensible and possibly even good :) (I don't really use these archiver and pruner functions in _my_ workflow so I am mostly :man_shrugging:  here) but the _process of dropping them unannounced_ is not one I prefer in my repos.  That was all.

**Comment by eddelbuettel** [comment]
I will let this setlle for a bit and ponder, but it may now be ready for merge, maybe by tomorrow.  Having an actual test is step up in rigour so thanks for putting it together!

File: NAMESPACE (hunk 20 above)
Comment (FelixErnst) [reply NAMESPACE:11]: Totally agree, However, I didn't drop anything, just slotted the additional functions in there at there appropriate places.

File: DESCRIPTION (hunk 19 above)
Comment (FelixErnst) [reply DESCRIPTION:9]: I totally agree. However, I didn't want to highjack Patrick's idea. It was his and he did the hard work of coming up with it.

**Comment by pat-s** [comment]
I haven't had time to follow up here.
Thanks for everyone pushing this forward.

AFAICS there is now a new `archivePackagesForAllRversions()`. Why is there an extra function needed to loop over different R versions? This is a quite unusual design decision.

In addition, I tried `archivePackagesForAllRversions()` on the example repo of the first post in this PR and the macOS binaries for R 4.0 did not get archived. @FelixErnst Could you have a look? Setting up the repo should only take a few seconds (see the snippet in the first post).

**Comment by FelixErnst** [comment]
In a week when I am back
________________________________

**Comment by FelixErnst** [comment]
So I looked into this

- `bin/macosx/mac.binary/contrib/4.0` is not a path valid for R `4.0` packages to be installed from. `bin/macosx/contrib/4.0` is the correct path. So I guess your repo is broken at least for the mac binaries.
- `archivePackages` used to work on packages of the current R version (current as in R version the user did run `drat` from). This changed by introducing the version argument in `0.1.6` and by the changes you introduced for vectorization. (You specified `version = NULL`, which I changed to `version = NA` to distinguish `no version == all versions` (`NA`), `current version` (`NULL`) and specific version.) By having `version = NA` as default, the user can very easily work on all folders of the repo, but there is a danger, that a surgical removal of packages for a specific version is not available or always overwritten by default. For your conveniance and a clear behavior I added the `archivePackagesForAllRversions()` function. The name is very bulky, but I couldn't think of an alternative.

**Comment by pat-s** [comment]
Thanks for taking the time.

> bin/macosx/mac.binary/contrib/4.0 is not a path valid for R 4.0 packages to be installed from. bin/macosx/contrib/4.0 is the correct path. So I guess your repo is broken at least for the mac binaries.

The current repo state is ok but you are right, it had the wrong paths for the revision given in the first post.

Seems to work as expected.
